### PR TITLE
Lecture 09 (tarab / Tara Boroushaki)

### PR DESCRIPTION
**Pull request by TaraBrshk**

### CODE REVIEW

File: notes/lecture-09-tarab.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
+# Lecture 09:  Neural Architecture Search Part III
+
+## Note Information
+
+| Title       | Neural Architecture Search Part III                                                                                            |
+|-------------|-----------------------------------------------------------------------------------------------------------------|
+| Lecturer    | Song Han \& Lucas Liebenwein                                                                                                       |
+| Date        | 10/06/2022                                                                                                      |
+| Note Author | Tara Boroushaki (tarab)                                                                                      |
+| Description | In this lecture, OMNIMIZER, a tool for neural network hardware-aware optimization, was introduced.
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:10]:
```suggestion
| Description | In this lecture, Omnimizer, a tool for neural network hardware-aware optimization, was introduced. 
```

File: notes/lecture-09-tarab.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
+# Lecture 09:  Neural Architecture Search Part III
+
+## Note Information
+
+| Title       | Neural Architecture Search Part III                                                                                            |
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:5]:
```suggestion
| Title       | Neural Architecture Search (Part III)                                                                                            |
```

File: notes/lecture-09-tarab.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
+# Lecture 09:  Neural Architecture Search Part III
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:1]:
```suggestion
# Lecture 09:  Neural Architecture Search (Part III)
```

File: notes/lecture-09-tarab.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+
+ # Conclusion:
+ 
+**OMNIMIZER NAS**:
+ - Pytorch-native model optimization and adaptation
+ - Can do both AutoNAS and FastNAS
+ - Searches for optimal subnet
+
+**OMNIMIZER Engine**:
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:120]:
```suggestion
**Omnimizer Engine**:
```

File: notes/lecture-09-tarab.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+```
+from omnimizer import engine device_model = engine.compile(sub_model)
+device_model.get_latency()
+out = device_model(sample_input)
+ ```
+
+
+ # Conclusion:
+ 
+**OMNIMIZER NAS**:
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:115]:
```suggestion
**Omnimizer NAS**:
```

File: notes/lecture-09-tarab.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    model,
+    constraints,
+    deployment,
+)
+ ```
+
+ ### Deploy:
+
+```
+from omnimizer import engine device_model = engine.compile(sub_model)
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:107]:
```suggestion
from omnimizer import engine
device_model = engine.compile(sub_model)
```

File: notes/lecture-09-tarab.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    "precision": "int8",
+}
+engine.get_latency(model, deployment)
+```
+
+### Diagnosis & Adaptation:
+The adaptation methods are based on real hardware deployments.
+```
+adapted_model = nas.adapt(model, deployment) 
+engine.profill(adapted_model, deployment)
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:75]:
```suggestion
engine.profile(adapted_model, deployment)
```

File: notes/lecture-09-tarab.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+- Autonomous driving and Advanced driver assistance systems 
+- Robotics
+- Internet of 
+- Augmented Reality and Virtual Reality Headsets
+
+### What is the problem that OMNIMIZER is trying to solve?
+There is a big gap between training neural networks and hardware deployment. It takes a long time to deploy AI models on actual hardwares. Even after deployment, there is a significant gap between expected performance and how the AI performs on the hardware. With current setups, a huge R&D cost and investment is always needed.
+
+
+## OMNIMIZER workflow:
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:30]:
```suggestion
## Omnimizer workflow:
```

File: notes/lecture-09-tarab.md [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+- Accelerate design and developement of neural networks
+- Give developers control through self-service tools
+- Faster Model inference 
+- Getting the best of current available hardware 
+
+
+### Edge AI applications:
+- Autonomous driving and Advanced driver assistance systems 
+- Robotics
+- Internet of
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:23]:
```suggestion
- Internet of Things (IoT)
```

File: notes/lecture-09-tarab.md [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+- Getting the best of current available hardware 
+
+
+### Edge AI applications:
+- Autonomous driving and Advanced driver assistance systems 
+- Robotics
+- Internet of 
+- Augmented Reality and Virtual Reality Headsets
+
+### What is the problem that OMNIMIZER is trying to solve?
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:26]:
```suggestion
### What is the problem that Omnimizer is trying to solve?
```

File: notes/lecture-09-tarab.md [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+## OMNIMIZER workflow:
+
+#### Current workflow:
+
+The ML engineers develop and train a network, providing the model to the deployment engineers who implement the network on the intended hardware. The ML engineers have to wait to get feedback from the deployment engineer on the latency, memory, and energy consumption and change the network to match the requirement of the intended product.
+This process is very time consuming.
+
+#### How OMNIMIZER changes the overall workflow:
+OMNIMIZER provides the ML engineer with a platform to deploy the network on an actual hardware and get the latency, memory, and energy consumption estimation. This capability shorten the R\&D cost and investment and reduces the time needed to build new products.
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:38]:
```suggestion
Omnimizer provides the ML engineer with a platform to deploy the network on an actual hardware and get the latency, memory, and energy consumption estimation. This capability shorten the R\&D cost and investment and reduces the time needed to build new products.
```

File: notes/lecture-09-tarab.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+OMNIMIZER can be added to Pytorch and enables The ML engineer to deploy and test on target hardware through Omnimizer Core and Engine.
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:40]:
```suggestion
Omnimizer can be added to Pytorch and enables The ML engineer to deploy and test on target hardware through Omnimizer Core and Engine.
```

File: notes/lecture-09-tarab.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+ 
+```
+
+### Train and Optimize:
+Uses FastNAS (Faster but smaller search space) or AutoNAS (Slower but larger search space).
+```
+from omnimizer import nas
+omni_model = nas.omnimize(model)
+train(omni_model, dataloader)
+
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:85]:
```suggestion
```

File: notes/lecture-09-tarab.md [hunk 14]
```diff
@@ -0,0 +1,124 @@
+# Lecture 09:  Neural Architecture Search Part III
+
+## Note Information
+
+| Title       | Neural Architecture Search Part III                                                                                            |
+|-------------|-----------------------------------------------------------------------------------------------------------------|
+| Lecturer    | Song Han \& Lucas Liebenwein                                                                                                       |
```
Comment (lucaslie) [inline notes/lecture-09-tarab.md:7]: please add my email: lucasl@omniml.ai

File: notes/lecture-09-tarab.md (hunk 12 above)
Comment (lucaslie) [inline notes/lecture-09-tarab.md:40]:
```suggestion
OMNIMIZER can be added to Pytorch and enables the ML engineer to deploy and test on target hardware through Omnimizer Core and Engine.
```

Comment (lucaslie) [inline notes/lecture-09-tarab.md:40]: Please explain "Core" and "Engine" in a little more detail (1 sentence each).

File: notes/lecture-09-tarab.md [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+## What are other approaches and competitors?
+
+### Microsoft (Neural Network Intelligence): 
+- Has a collection of algorithm
+- Needs manual search space design 
+- Works good only on well-known benchmarks such as ImageNet
+
+### Google (Vertex AI):
+- A black box design
+- Works good only on well-known benchmarks such as ImageNet
```
Comment (lucaslie) [inline notes/lecture-09-tarab.md:51]:
```suggestion
- Works well only on established benchmarks such as ImageNet
```

File: notes/lecture-09-tarab.md [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+### Omnimizer: 
+- Automatic optimization and adaptation
+- Can work with any training environment 
+- Instant deployment on hardware (that are supported)
+
+
+## Omnimizer Workflow:
+### Setup:
+OmniML has several actual hardwares that their server deploys and measures latency.
```
Comment (lucaslie) [inline notes/lecture-09-tarab.md:61]: This doesn't very clear. Please try to explain it a bit more.

File: notes/lecture-09-tarab.md [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+from omnimizer import engine
+deployment = {
+    "device": "S888",
+    "precision": "int8",
+}
+engine.get_latency(model, deployment)
+```
+
+### Diagnosis & Adaptation:
+The adaptation methods are based on real hardware deployments.
```
Comment (lucaslie) [inline notes/lecture-09-tarab.md:72]: same here.

**Review by lucaslie — COMMENTED** [review]
Looks good overall. Left a few comments.

**Comment by zhijian-liu** [comment]
@TaraBrshk, please let us know when the scribe is ready for review. Thanks!

File: notes/lecture-09-tarab.md [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+```
+adapted_model = nas.adapt(model, deployment) 
+engine.profile(adapted_model, deployment)
+ 
+```
+
+### Train and Optimize:
+Uses FastNAS (Faster but smaller search space) or AutoNAS (Slower but larger search space).
+```
+from Omnimizer import nas
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:82]:
```suggestion
from omnimizer import nas
```

File: notes/lecture-09-tarab.md [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+- Automatic optimization and adaptation
+- Can work with any training environment 
+- Instant deployment on hardware (that are supported)
+
+
+## Omnimizer Workflow:
+### Setup:
+OmniML has several actual hardwares that their server deploys and measures latency.
+```
+from Omnimizer import engine
```
Comment (zhijian-liu) [inline notes/lecture-09-tarab.md:63]:
```suggestion
from omnimizer import engine
```

**Comment by zhijian-liu** [comment]
@TaraBrshk, thanks for the update! Please fix the remaining comments above and re-request the review. Btw, please also remove `.DS_Store`.

**Review by zhijian-liu — APPROVED** [review]
Thanks for your contribution.

**Pull request merged**